### DISCUSSION TITLE
using qemu

### DISCUSSION THREAD

**Discussion by djdisodo** [Rahix/avr-hal#280] [Show and tell]
in this case you only need `qemu-system-avr` and `avr-gcc` other than rust

both are available in most linux distros and msys2-mingw(for windows)

add this to your `.cargo/config.toml` of your project
```
[target.'cfg(target_arch = "avr")']
runner = "qemu-system-avr -M uno -nographic -serial stdio -bios"
```

uno is board i am using
here's list of boards supported by qemu
```
$ qemu-system-avr -machine help
Supported machines are:
2009                 Arduino Duemilanove (ATmega168) (alias of arduino-duemilanove)
arduino-duemilanove  Arduino Duemilanove (ATmega168)
mega2560             Arduino Mega 2560 (ATmega2560) (alias of arduino-mega-2560-v3)
arduino-mega-2560-v3 Arduino Mega 2560 (ATmega2560)
mega                 Arduino Mega (ATmega1280) (alias of arduino-mega)
arduino-mega         Arduino Mega (ATmega1280)
uno                  Arduino UNO (ATmega328P) (alias of arduino-uno)
arduino-uno          Arduino UNO (ATmega328P)
none                 empty machine
```

```
[target.'cfg(target_arch = "avr")']
runner = "qemu-system-avr -M uno -bios"
```

without option nographic and serial there will be qemu console
you can use serial console on view->serial0

this vm may run faster than real hw
current implementation of delay won't work properly

can we have a wiki?

**Comment by Rahix** [comment]
Sounds good, I'll enable the wiki.
